state/state.md

```diff
@@ -1 +1,45 @@
-# State
+# BlockExecutor
+
+## Abstract
+
+ The `BlockExecutor` is a component responsible for creating, applying, and maintaining blocks and state in the system. It interacts with the mempool and the application via the ABCI interface.
+
+## Component Description
+
+ The `BlockExecutor` is initialized with a proposer address, `namespace ID`, `chain ID`, `mempool`, `proxyApp`, `eventBus`, and `logger`. It uses these to manage the creation and application of blocks. It also validates blocks and commits them, updating the state as necessary.
+
+## Detailed Description
+
+- `NewBlockExecutor`: This method creates a new instance of `BlockExecutor`. It takes a proposer address, `namespace ID`, `chain ID`, `mempool`, `proxyApp`, `eventBus`, and `logger` as parameters.
+- `InitChain`: This method initializes the chain by calling `InitChainSync` using the consensus connection to the app. It takes a `GenesisDoc` as a parameter.
+- `CreateBlock`: This method reaps transactions from the mempool and builds a block. It takes the current state, the height of the block, last header hash, and the commit as parameters.
+- `ApplyBlock`: This method applies the block to the state. It takes the current state, the block ID, and the block itself as parameters. It can return the following named errors:
+  - `ErrEmptyValSetGenerate`: returned when applying the validator changes would result in empty set.
+  - `ErrAddingValidatorToBased`: returned when adding validators to empty validator set.
+- `Validate`: This method validates the block. It takes the state and the block as parameters. It applies the following validations:
+  - New block version must match state block version.
+  - New block height must match state initial height.
+  - New block height must be adjacent to state last block height.
+  - New block header `AppHash` must match state `AppHash`.
+  - New block header `LastResultsHash` must match state `LastResultsHash`.
+  - New block header `AggregatorsHash` must match state `Validators.Hash()`.
+- `commit`: This method commits the block and updates the state. It takes the state, the block, and a list of ABCI `ResponseDeliverTx` as parameters.
+- `execute`: This method executes the block. It takes the context, the state, and the block as parameters.
```

Comment:
Can we expand on what execute here means? Something referring to calling the relevant ABCI methods like `FinalizeBlock`

Reply:
Added a sentence to mention `BeginBlock`, `DeliverTx`, `EndBlock`. I don't see any references to `FinalizeBlock`, I guess not yet being used.

Reply:
We should be using the `v0.11.x` branch which has `FinalizeBlock`: https://github.com/rollkit/rollkit/blob/05ff97ea78862631436c8b638d50a63de0e1b7b0/state/executor.go#L301